src/AMSlib/wf/resource_manager.cpp

```diff
@@ -5,48 +5,113 @@
  * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
  */
 
-#include "resource_manager.hpp"
-
+#include <cstdlib>
+#include <cstring>
 #include <umpire/ResourceManager.hpp>
 #include <umpire/Umpire.hpp>
 #include <umpire/strategy/QuickPool.hpp>
 
 #include "debug.h"
+#include "device.hpp"
+#include "resource_manager.hpp"
 
 namespace ams
 {
 
-std::string AMSAllocator::getName() { return allocator.getName(); }
-
-void *AMSAllocator::allocate(size_t num_bytes)
+template<typename T>
+static T roundUp(T num_to_round, int multiple) 
 {
-  void *ptr = allocator.allocate(num_bytes);
-  CFATAL(ResourceManager,
-         ptr == nullptr,
-         "Failed to allocated %ld values using allocator %s",
-         num_bytes,
-         getName().c_str());
-  return ptr;
+    return ((num_to_round + multiple- 1) / multiple) * multiple;
 }
 
-void AMSAllocator::deallocate(void *ptr) { allocator.deallocate(ptr); }
+std::string AMSAllocator::getName() { return name; }
+
+
+struct AMSDefaultDeviceAllocator final : AMSAllocator {
+  AMSDefaultDeviceAllocator(std::string name) : AMSAllocator(name){};
+  ~AMSDefaultDeviceAllocator() = default;
```

Comment:
Shouldn't it also free/deallocate allocations in the destructor?

Reply:
We are not keeping track of allocations. It is the responsibility of the caller to deallocate data.